File: backend/app/models/post.py
from datetime import datetime, timezone
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON, Enum as SAEnum
from sqlalchemy.orm import relationship
import enum
from app.database import Base


class ProcessingStatus(str, enum.Enum):
    pending = "pending"
    processing = "processing"
    completed = "completed"
    failed = "failed"


class ApprovalStatus(str, enum.Enum):
    pending_review = "pending_review"  # freshly scraped, awaiting admin decision
    approved = "approved"              # admin approved — eligible for AI processing
    rejected = "rejected"              # admin rejected — will not be processed


class ScrapedPost(Base):
    """Raw post data extracted from a LinkedIn source."""

    __tablename__ = "scraped_posts"

    id = Column(Integer, primary_key=True, index=True)
    source_id = Column(Integer, ForeignKey("sources.id", ondelete="CASCADE"), nullable=False, index=True)
    post_link = Column(String, unique=True, nullable=False, index=True)
    post_text = Column(Text, nullable=True)
    image_url = Column(String, nullable=True)
    timestamp = Column(DateTime(timezone=True), nullable=True)
    approval_status = Column(
        SAEnum(ApprovalStatus),
        default=ApprovalStatus.pending_review,
        nullable=False,
        index=True,
    )
    reviewed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    source = relationship("Source", backref="scraped_posts")
    processed = relationship("ProcessedPost", back_populates="scraped_post", uselist=False)


class ProcessedPost(Base):
    """AI-rewritten and image-generated version of a scraped post."""

    __tablename__ = "processed_posts"

    id = Column(Integer, primary_key=True, index=True)
    scraped_post_id = Column(Integer, ForeignKey("scraped_posts.id", ondelete="SET NULL"), unique=True, nullable=True, index=True)
    source_id = Column(Integer, nullable=True, index=True)   # denormalized – survives source deletion
    source_label = Column(String, nullable=True)              # denormalized – survives source deletion
    rewritten_post = Column(Text, nullable=True)
    hooks = Column(Text, nullable=True)
    hashtags = Column(JSON, nullable=True)  # list of strings
    generated_image_url = Column(String, nullable=True)
    status = Column(SAEnum(ProcessingStatus), default=ProcessingStatus.pending, nullable=False, index=True)
    error_message = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    scraped_post = relationship("ScrapedPost", back_populates="processed")

File: frontend/types/post.ts
export type ApprovalStatus = "pending_review" | "approved" | "rejected";
export type ProcessingStatus = "pending" | "processing" | "completed" | "failed";
export type ReimagineStatus = "idle" | "generating";

export interface ScrapedPost {
  id: number;
  source_id: number;
  post_link: string;
  post_text: string | null;
  image_url: string | null;
  timestamp: string | null;
  approval_status: ApprovalStatus;
  reviewed_at: string | null;
  created_at: string;
}

export interface ImageVersion {
  id: number;
  image_path: string;
  created_at: string;
}

export interface ProcessedPost {
  id: number;
  scraped_post_id: number | null;
  source_id: number | null;
  source_label: string | null;
  rewritten_post: string | null;
  hooks: string | null;
  hashtags: string[] | null;
  generated_image_url: string | null;
  reimagine_status: ReimagineStatus;
  status: ProcessingStatus;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface FullPost {
  scraped_post_id: number;
  source_id: number;
  post_link: string;
  original_text: string | null;
  original_image_url: string | null;
  post_timestamp: string | null;
  approval_status: ApprovalStatus;
  processed_post_id: number | null;
  rewritten_post: string | null;
  hooks: string | null;
  hashtags: string[] | null;
  generated_image_url: string | null;
  download_image_url: string | null;
  copy_ready_text: string | null;
  reimagine_status: ReimagineStatus;
  image_versions: ImageVersion[];
  status: ProcessingStatus | "unprocessed";
}
